internal/sshtunneler/tunneler.go

```diff
@@ -0,0 +1,263 @@
+// Copyright 2025 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package sshtunneler
+
+import (
+	"context"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/juju/clock"
+	"github.com/juju/errors"
+	"github.com/lestrrat-go/jwx/v2/jwa"
+	gossh "golang.org/x/crypto/ssh"
+
+	"github.com/juju/juju/core/network"
+	"github.com/juju/juju/pki/ssh"
+	"github.com/juju/juju/state"
+)
+
+const (
+	maxTimeout        = 60 * time.Second
+	reverseTunnelUser = "reverse-tunnel"
+	tokenIssuer       = "sshtunneler"
+	tokenSubject      = "reverse-tunnel"
+	tunnelIDClaim     = "tunnelID"
+	defaultUser       = "ubuntu"
+)
+
+// State writes requests for tunnels to state.
```

Comment:
State defines an interface used by the TunnelTracker to write tunnel requests to state.